scarlet/update.py

```diff
@@ -214,13 +216,115 @@ def translation(component, direction=1, kernel=interpolation.lanczos, padding=3)
+    # off the edge of the psf array
+    y_morph_ext = np.arange(morph.shape[0])
+    x_morph_ext = np.arange(morph.shape[1])
+
+    # calculate the offset in the morph frame such that the center pixel aligns
+    # with the psf coordindate frame, where the psf peak is in the center
+    y_morph_ext = y_morph_ext - cy
+    x_morph_ext = x_morph_ext - cx
+
+    # Compare the two end points, and take whichever is the smaller radius
+    # use that to select the entries that are equal to or below that radius
+    # Find the minimum between the end points (if the psf goes off the edge,
+    # and the radius of the psf
+    rad_endpoints_y = np.min((abs(y_morph_ext[0]), abs(y_morph_ext[-1]), y_rad))
+    rad_endpoints_x = np.min((abs(x_morph_ext[0]), abs(x_morph_ext[-1]), x_rad))
+    trimmed_y_range = y_morph_ext[abs(y_morph_ext) <= rad_endpoints_y]
+    trimmed_x_range = x_morph_ext[abs(x_morph_ext) <= rad_endpoints_x]
+
+    psf_y_range = trimmed_y_range + psf_peak_y
+    psf_x_range = trimmed_x_range + psf_peak_x
+
+    morph_y_range = trimmed_y_range + cy
+    morph_x_range = trimmed_x_range + cx
+
+    # Use these arrays to get views of the morphology and psf
+    morph_view = morph[morph_y_range][:, morph_x_range]
+    psf_view = psf[psf_y_range][:, psf_x_range]
+
+    morph_view_weighted = morph_view*psf_view
+    morph_view_weighted_sum = np.sum(morph_view_weighted)
+    # build the indices to use the the centroid calculation
+    indy, indx = np.indices(psf_view.shape)
+    first_moment_y = np.sum(indy*morph_view_weighted) / morph_view_weighted_sum
+    first_moment_x = np.sum(indx*morph_view_weighted) / morph_view_weighted_sum
+    # build the offset to go from psf_view frame to psf frame to morph frame
+    # aka move the peak back by the radius of the psf width adusted for the
+    # minimum point in the view
+    offset = (morph_y_range[0], morph_x_range[0])
+
+    whole_pixel_center = np.round((first_moment_y, first_moment_x))
+    dy, dx = whole_pixel_center - (first_moment_y, first_moment_x)
+    morph_pixel_center = tuple((whole_pixel_center + offset).astype(int))
+    component.pixel_center = morph_pixel_center
+    component.shift = (dy, dx)
+    return component
+
+
+def symmetric(component, algorithm="kspace", bbox=None, fill=None, strength=.5):
     """Make the source symmetric about its center
 
     See `~scarlet.operator.prox_uncentered_symmetry`
     for a description of the parameters.
     """
+    pixel_center = component.pixel_center
+    if bbox is not None:
+        # Only apply monotonicity to the pixels inside the bounding box
+        morph = component.morph[bbox.slices]
+        shape = morph.shape
+        if shape[0] <= 1 or shape[1] <= 1:
+            return component
+        center = pixel_center[0]-bbox.bottom, pixel_center[1]-bbox.left
+    else:
+        morph = component.morph
+        shape = component.shape[-2:]
+        center = pixel_center
+    morph = morph.copy()
```

Comment:
Why is there a copy here? Isn't the point of an update to overwrite the original?

Reply:
I just copied the code from `monotonicity` that used the bounding box without thinking. I don't remember why we needed it there, I think it might have had something to do with trying to use proximal operators with `autograd` arrays when I first started out with them and there being some sort of conversion issue between `autograd` arrays and `Eigen` matrices that we use in monotonicity (`autograd.numpy.Array.copy` returns a numpy array, not an array with gradients... I think). I'll fix this and push the commit.